firedrake/bcs.py

```diff
@@ -319,6 +319,8 @@ def reconstruct(self, field=None, V=None, g=None, sub_domain=None, method=None,
            V.parent == fs.parent and \
            V.index == fs.index and \
            V.component == fs.component and \
+           (V.parent is None or V.parent.parent == fs.parent.parent) and \
+           (V.parent is None or V.parent.index == fs.parent.index) and \
```

Comment:
Thinking about this, this is a more general statement about the equality of function spaces, no?

Should we not be writing this as:
```python
if V == fs and g == ...:
   return self
```
And implement `V == fs` correctly on functionspaces?

Reply:
So, `V.component == ...` is nicely removed from here as it has actually been in `FunctionSpace.__eq__` already.
Moving `V.parent ==` causes recursions (`FunctionSpace.__eq__` <->`MixedFunctionSpace.__eq__` ).
Shouldn't `V` still be regarded as equal to `V` even if it is embedded in some `MixedFunctionSpace`, `W`, so that `V == W.sub(0)` holds?
I think the reason we have to check `parents` here is that `DirichletBC._indices` has to be consistent to whether we have `parent` or `parent.parent`, so isn't it fair to leave these checks here?

Reply:
OK, makes sense.